### PR TITLE
Added copy with modification to DimensionRow and build a ResultSetMap…

### PR DESCRIPTION
**Pull request by michael-mclawhorn**
…per able to transform dimension rows from a ResultSet.

### CODE REVIEW

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/DimensionRow.java [hunk 1]
```diff
@@ -63,4 +77,18 @@ public int compareTo(DimensionRow that) {
         return entrySet().stream()
                 .collect(StreamUtils.toLinkedMap(entry -> entry.getKey().getName(), Map.Entry::getValue));
     }
+
+    /**
+     * Create a copy of a dimension row, and then transform it using a mapping function.
```
Comment (archolewa) [inline fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/DimensionRow.java]: It isn't really clear from the documentation that when we mean "transform" it we mean, apply a mapping to each dimension field's value. Something like:

```
Copies a DimensionRow and transforms its fields using the specified mapper.

@param row The row to be transformed
@param mapper  A function that takes a DimensionField and its current value, and returns the field's new value.

@return A copy of the DimensionRow with its fields transformed by the specified function
```

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/metric/mappers/MemoizingDimensionMappingResultSetMapper.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+ * A ResultSetMapper that transforms a ResultSet by transforming its {@link DimensionRow}s.
+ *
+ * If memoization is used, transformed rows will be cached.  The cost of memoization is proportional to the hashcode
+ * cost of the DimensionRow entries and the value is proportional to the number of repeated rows and the cost of
+ * transforming them.  In general, the expectation is for high redundancy of dimension rows per result set, so
+ * memoization is on by default.
+ */
+public class MemoizingDimensionMappingResultSetMapper extends ResultSetMapper {
+
+    private final BiFunction<DimensionColumn, DimensionRow, Boolean> columnRowMatcher;
```
Comment (archolewa) [inline fili-core/src/main/java/com/yahoo/bard/webservice/data/metric/mappers/MemoizingDimensionMappingResultSetMapper.java]: `java.util.function` has a `BiPredicate`: https://docs.oracle.com/javase/8/docs/api/java/util/function/BiPredicate.html that you could use. This should simplify this type some, and make it clearer what this is doing.

File: fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/DimensionRowSpec.groovy [hunk 3]
```diff
@@ -26,4 +29,31 @@ class DimensionRowSpec extends Specification {
...
+
+    def "missing key throws error"() {
+        when:
+        DimensionRow testRow = new DimensionRow(missingKeyField, [(keyField): keyValue, (nonKeyField): nonKeyValue])
+
+        then:
+        thrown(IllegalArgumentException)
+    }
+
+    def "copyWithReplace copies and replaces"() {
```
Comment (archolewa) [inline fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/DimensionRowSpec.groovy]: The name implies that it copies and replaces, but I don't see any validation that it actually copies (i.e. leaves the original unmodified). Should we include an additional assertion that `testRow != expectedRow`?

File: fili-core/src/test/groovy/com/yahoo/bard/webservice/web/endpoints/DimensionsServletComponentSpec.groovy [hunk 4]
```diff
@@ -321,7 +321,7 @@ class DimensionsServletComponentSpec extends Specification {
         result.head().getHeaderString("Content-Disposition") == "attachment; filename=dimensions-shape-values.csv"
         result.head().getHeaderString("Content-Type") == "text/csv; charset=utf-8"
         Map<String, String> actualLinks = GroovyTestUtils.splitHeaderLinks(result.head().getHeaderString("Link"))
-        actualLinks.every {name, actualLink -> GroovyTestUtils.compareURL(actualLink, expectedLinks[name])}
+        actualLinks.every {name, actualLink -> GroovyTestUtils.compareURL(actualLink, expectedLinks [name])}
```
Comment (archolewa) [inline fili-core/src/test/groovy/com/yahoo/bard/webservice/web/endpoints/DimensionsServletComponentSpec.groovy]: Any particular reason for this change?

/NonBlocker

Comment (michael-mclawhorn) [reply fili-core/src/test/groovy/com/yahoo/bard/webservice/web/endpoints/DimensionsServletComponentSpec.groovy]: it actually got in accidentally.
I'll remove it.

File: fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/DimensionRowSpec.groovy (hunk 3 above)
Comment (michael-mclawhorn) [reply fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/DimensionRowSpec.groovy]: done

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/metric/mappers/MemoizingDimensionMappingResultSetMapper.java (hunk 2 above)
Comment (michael-mclawhorn) [reply fili-core/src/main/java/com/yahoo/bard/webservice/data/metric/mappers/MemoizingDimensionMappingResultSetMapper.java]: done

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/DimensionRow.java (hunk 1 above)
Comment (michael-mclawhorn) [reply fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/DimensionRow.java]: indeed, clearer.

**Pull request merged**